x/gamm/twap/store.go

```diff
@@ -15,17 +24,121 @@ func (k Keeper) trackChangedPool(ctx sdk.Context, poolId uint64) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FormatMostRecentTWAPKey(poolId, asset0Denom, asset1Denom)
+	bz := store.Get(key)
+	return types.ParseTwapFromBz(bz)
+}
+
+// getMostRecentRecordStoreRepresentation returns all most recent twap records
+// (in state representation) for the provided pool id.
+func (k Keeper) getAllMostRecentRecordsForPool(ctx sdk.Context, poolId uint64) ([]types.TwapRecord, error) {
+	store := ctx.KVStore(k.storeKey)
+	return types.GetAllMostRecentTwapsForPool(store, poolId)
+}
+
+// storeNewRecord stores a record, in both the most recent record store and historical stores.
+func (k Keeper) storeNewRecord(ctx sdk.Context, twap types.TwapRecord) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FormatMostRecentTWAPKey(twap.PoolId, twap.Asset0Denom, twap.Asset1Denom)
+	osmoutils.MustSet(store, key, &twap)
+	k.storeHistoricalTWAP(ctx, twap)
+}
+
+// getRecordAtOrBeforeTime on a given input (id, t, asset0, asset1)
+// returns the TWAP record from state for (id, t', asset0, asset1),
+// where t' is such that:
+// * t' <= t
+// * there exists no t'' in state, where t'' < t'
+//
+// This returns an error if:
+// * there is no historical record in state at or before t
+//   - Occurs if t is older than pruning period, or pool creation date.
+// * there is no record for the asset pair (asset0, asset1) in particular
+//   - e.g. asset not in pool, or provided in wrong order.
+func (k Keeper) getRecordAtOrBeforeTime(ctx sdk.Context, poolId uint64, t time.Time, asset0Denom string, asset1Denom string) (types.TwapRecord, error) {
+	store := ctx.KVStore(k.storeKey)
+	// We make an iteration from time=t + 1ns, to time=0 for this pool.
+	// Note that we cannot get any time entries from t + 1ns, as the key would be `prefix|t+1ns`
+	// and the end for a reverse iterator is exclusive. Thus the largest key that can be returned
+	// begins with a prefix of `prefix|t`
+	startKey := types.FormatHistoricalPoolIndexTimePrefix(poolId, time.Unix(0, 0))
+	endKey := types.FormatHistoricalPoolIndexTimePrefix(poolId, t.Add(time.Nanosecond))
+	lastParsedTime := time.Time{}
+	stopFn := func(key []byte) bool {
+		// halt iteration if we can't parse the time, or we've successfully parsed
+		// a time, and have iterated beyond records for that time.
+		parsedTime, err := types.ParseTimeFromHistoricalPoolIndexKey(key)
+		if err != nil {
+			return true
+		}
+		if lastParsedTime.After(parsedTime) {
+			return true
+		}
+		lastParsedTime = parsedTime
+		return false
+	}
+
+	reverseIterate := true
+	twaps, err := osmoutils.GetIterValuesWithStop(store, startKey, endKey, reverseIterate, stopFn, types.ParseTwapFromBz)
+	if err != nil {
+		return types.TwapRecord{}, err
+	}
+	if len(twaps) == 0 {
+		return types.TwapRecord{}, errors.New("looking for a time thats too old, not in the historical index. " +
+			" Try storing the accumulator value.")
+	}
+
+	for _, twap := range twaps {
+		if twap.Asset0Denom == asset0Denom && twap.Asset1Denom == asset1Denom {
```

Comment:
Since denoms exist in a key, do we need this loop? 

We can probably make the `stopFn` stop once the right denoms are reached. Then, grab the last value

Reply:
This feels far simpler though

Reply:
Happy for us to explore the option your describing in a subsequent pass if you'd like

Reply:
Your explanation makes sense. As long as there is a good reason for this choice (and there is), I'm good with the approach :+1: